modules/library/data/src/main/java/org/geotools/data/memory/MemoryFeatureReader.java

```diff
@@ -30,20 +34,27 @@
 import org.opengis.feature.simple.SimpleFeatureType;
 
 public class MemoryFeatureReader implements FeatureReader<SimpleFeatureType, SimpleFeature>{
-    ContentState state;
+
     SimpleFeatureType featureType;
     Iterator<SimpleFeature> iterator;
 
     public MemoryFeatureReader(ContentState state, Query query) throws IOException {
-        this.state = state;
         featureType = state.getFeatureType();
-        String typeName = getFeatureType().getTypeName();
-        MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
-        iterator = store.features(typeName).values().iterator();
+
+        final MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
+        final Map<String, SimpleFeature> features = store.features(featureType.getTypeName());
+        synchronized (features) {
```

Comment:
Agree. I'm still evaluating how to refactor MemoryFeatureWriter because:
* Iterators are not synchronized if using Collections.synchronizedMap(), while access like put and get are --> Writer iterates to the end to add a new Feature ..
* implementing Iteration behavior and iterate to the feature + 1 on the next request could be a performance issue for writes  
* lack of knowledge of the workflow for Writers:
  * if iterator.hasNext() is false and next produces a new one
  * have the user to call write() before creating an other new one?
  * current Implementation looks like it would'nt work but I have to verifiy with a test
  * Ah! ContentFeatureStore.addFeatures(SimpleFeature, FeatureWriter) verifies my guessing, the user have to call write() after adding a feature

Where can I find some details about rendering behavior and event notifications on changes within a datastore. Normally if a User interacts with a map using UI its synchronized using UI-Thread. If the objects comming from a stream from another than the UI thread, how and when gets the map and its resources notified and re-rendered.

However I'm unsure about the current design for different reasons:
* the behavior is not really stateless (Iterator-Thing) 
* in addition it depends on a Query (what happens if the new features would match criterias for different Writers?)  

All in all I have a big Question Mark and looking forward for any feedback.